guide/avoiding-unnecessary-configuration.md

```diff
@@ -0,0 +1,161 @@
+```
+Any other lazy configuration of the task (eg in other plugins or scripts using `tasks).
+
+#### Good:
+
+Instead, we should use `register`:
+
+```java
+TaskProvider<FooTask> task = project.getTasks().register("foo", FooTask.class, foo -> {
+    foo.getInput().set(something);
+});
+```
+
+This is better; we tell Gradle there is task called `foo`, but we give the configuration of the task lazily. If `foo` is not executed, we don't even realise or configure the task.
+
+## Using `TaskProvider`s
+
+The `register` approach now gives us a `TaskProvider`, which is a lazy handle to the task. Many  of the built-in Gradle APIs will let you pass a `TaskProvider` instead of a `Task` object.
+
+If you want to give use an output property of a task as the input to another task, you can just transform the `TaskProvider` into another `Provider` using `map` or `flatMap`. You can call `.configure` on a `TaskProvider` to lazily configure the task without realising it (avoid calling `.get()` on `TaskProvider`s).
+
+```java
+abstract class MyTask extends DefaultTask {
+    @OutputFile
+    abstract RegularFileProperty getOutputFile();
+}
+```
+
+```java
+TaskProvider<MyTask> myTask = project.getTasks().register("myTask", MyTask.class);
+
+otherTaskProvider.configure(task -> {
+    task.getInput().set(myTask.flatMap(MyTask::getOutputFile));
+});
+```
+
+This also has the benefit of wiring up the dependencies between the two tasks automatically (no need for `dependsOn`).
+
+## Accessing tasks lazily
+
+> Further reading: [Task Configuration Avoidance: Eager APIs to avoid](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#eager_apis_to_avoid)
+
+To access tasks lazily, you need to make sure to call lazy APIs that return `TaskProvider` rather than `Task`.
+
+### Getting single tasks
+
+#### Bad:
+
+These all return a `Task` so will realise the task and run all the lazy configuration on it immediately:
+
+```java
+project.getTasks().getByName("foo");
+project.getTasks().findByName("foo");
+project.getTasks().getByPath(":foo");
+project.getTasks().findByPath(":foo");
+project.getTasks().withType(Foo.class).getByName(":foo");
+```
+
+#### Good:
+
+The `named` method returns a `TaskProvider`, allowing lazy configuration.
+
+```java
+TaskProvider<?> task = project.getTasks().named("foo");
+
+TaskProvider<FooTask> task = project.getTasks()
+        .withType(Foo.class)
+        .named("foo");
+```
+
+### Iterating over tasks
+
+#### Bad:
+
+These will immediately realise newly registered tasks when they are added (even if added using `register`) and run all the lazy configuration on them:
+
+```java
+project.getTasks().whenTaskAdded(task -> ...);
+project.getTasks().whenObjectAdded(task -> ...);
+```
+
+These will realise _all_ tasks, even ones already registered:
+
+```java
+project.getTasks().all(task -> ...);
+project.getTasks().matching(task -> ...);
+project.getTasks().withType(Foo.class).matching(task -> ...);
+```
```

Comment:
I think you also want to note the common pitfall of `withType(Foo.class, task -> {})` whilst they may look the same, this version is the equivalent of calling `withType().all()`. There are a number of different methods similar to this on the `TaskCollection`/`DomainObjectCollection` class hierarchy.